BigQuery/tests/System/RegionalizationTest.php

```diff
@@ -0,0 +1,207 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\BigQuery\Tests\System;
+
+use Google\Cloud\Core\Exception\NotFoundException;
+use Google\Cloud\Core\Exception\ServiceException;
+
+/**
+ * @group bigquery
+ * @group bigquery-regionalization
+ */
+class RegionalizationTest extends BigQueryTestCase
+{
+    const REGION_ASIA = 'asia-northeast1';
+    const REGION_US = 'US';
+    const QUERY_TEMPLATE = 'SELECT 1 FROM `%s.%s`';
+
+    private static $datasetAsia;
+    private static $tableAsia;
+
+    public static function setUpBeforeClass()
+    {
+        parent::setUpBeforeClass();
+        self::$datasetAsia = self::createDataset(
+            self::$client,
+            uniqid(self::TESTING_PREFIX),
+            ['location' => self::REGION_ASIA]
+        );
+        self::$tableAsia = self::createTable(
+            self::$datasetAsia,
+            uniqid(self::TESTING_PREFIX)
+        );
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testRunQuery($location)
+    {
+        $caught = false;
+        $query = self::$client->query(
+            sprintf(
+                self::QUERY_TEMPLATE,
+                self::$datasetAsia->id(),
+                self::$tableAsia->id()
+            )
+        )->location($location);
+
+        try {
+            self::$client->runQuery($query);
+        } catch (NotFoundException $e) {
+            $caught = true;
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testRunCopyJob($location)
+    {
+        $caught = false;
+        $copyTable = self::$dataset->table(self::TESTING_PREFIX);
+        $copy = self::$tableAsia->copy($copyTable)
+            ->location($location);
+
+        try {
+            self::$tableAsia->startJob($copy);
+            self::$deletionQueue->add($copyTable);
+        } catch (NotFoundException $e) {
+            $caught = true;
+        } catch (ServiceException $e) {
+            // Swallow this, as the important bit is whether or not
+            // we get a 404.
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testRunExtractJob($location)
+    {
+        $caught = false;
+        $extract = self::$tableAsia->extract(
+            self::$bucket->object(self::TESTING_PREFIX)
```

Comment:
I don't see where the bucket is created. You can only extract to a bucket that is in the same region, so I would want the test to be very clear that this bucket lives in the asia region..

Reply:
We should double check that the file exists in the bucket after a successful extract job run.